src/abstractdatatable/abstractdatatable.jl

```diff
@@ -706,83 +706,97 @@ Base.hcat(dt1::AbstractDataTable, dt2::AbstractDataTable) = hcat!(dt[:, :], dt2)
-                    if VERSION < v"0.5.0-dev" &&
-                       coltyps[idx] <: Nullable && !isa(coltyps[idx].types[2], DataType)
-                        coltyps[idx] = Nullable{Any}
-                    end
-                end
-                nonnull_ct[idx] += !_isnullable(col)
-            else # new column
-                push!(colindex, cn)
-                push!(coltyps, ct)
-                push!(similars, col)
-                push!(nonnull_ct, !_isnullable(col))
+"""
+    vcat(dts::AbstractDataTable...)
+
+Vertically concatenate `AbstractDataTables` that have the same column names in
+the same order.
+
+# Example
+```jldoctest
+julia> dt1 = DataTable(A=1:3, B=1:3);
+julia> dt2 = DataTable(A=4:6, B=4:6);
+julia> vcat(dt1, dt2)
+6×2 DataTables.DataTable
+│ Row │ A │ B │
+├─────┼───┼───┤
+│ 1   │ 1 │ 1 │
+│ 2   │ 2 │ 2 │
+│ 3   │ 3 │ 3 │
+│ 4   │ 4 │ 4 │
+│ 5   │ 5 │ 5 │
+│ 6   │ 6 │ 6 │
+```
+"""
+Base.vcat(dt::AbstractDataTable) = dt
+function Base.vcat(dts::AbstractDataTable...)
+    isempty(dts) && return DataTable()
+    allheaders = map(names, dts)
+    if all(h -> length(h) == 0, allheaders)
+        return DataTable()
+    end
+    uniqueheaders = unique(allheaders)
+    if length(uniqueheaders) > 1
+        unionunique = union(uniqueheaders...)
+        coldiff = setdiff(unionunique, intersect(uniqueheaders...))
+        if !isempty(coldiff)
+            # if any datatables are a full superset of names, skip them
+            filter!(u -> Set(u) != Set(unionunique), uniqueheaders)
+            estrings = Vector{String}(length(uniqueheaders))
+            for (i, u) in enumerate(uniqueheaders)
+                matching = find(h -> u == h, allheaders)
+                headerdiff = setdiff(coldiff, u)
+                cols = join(headerdiff, ", ", " and ")
+                args = join(matching, ", ", " and ")
+                estrings[i] = "column(s) $cols are missing from argument(s) $args"
             end
+            throw(ArgumentError(join(estrings, ", ", ", and ")))
+        else
+            estrings = Vector{String}(length(uniqueheaders))
+            for (i, u) in enumerate(uniqueheaders)
+                indices = find(a -> a == u, allheaders)
+                estrings[i] = "column order of argument(s) $(join(indices, ", ", " and "))"
+            end
+            throw(ArgumentError(join(estrings, " != ")))
         end
-    end
-
-    for j in 1:length(colindex)
-        if nonnull_ct[j] < length(dts) && !_isnullable(similars[j])
-            similars[j] = EMPTY_DATA
+    else
+        header = uniqueheaders[1]
+        cols = Vector{Any}(length(header))
+        for i in 1:length(cols)
+            data = [dt[i] for dt in dts]
+            lens = map(length, data)
+            cols[i] = promote_col_type(data...)(reduce(+, lens))
+            copy!(cols[i], data[1])
+            for j in 2:length(data)
+                offset = reduce(+, lens[1:j-1]) + 1
```

Comment:
`offset = lens[1]` for initialization and `offset += lens[j]` after each iteration would be simpler.